In [ ]:
# Imports and pip installations (if needed)
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier

from sklearn import datasets
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score


# Part 1: Load the dataset

In [3]:
# Load the dataset (load remotely or locally)
iris_df = pd.read_csv('iris.csv')
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
print(iris_df.head(15))

iris_df.info()

    sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14           5.8         

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

In our data we have columns named Sepal_length, Sepal_width, Petal_Length, Petal_width, and Species. The first 4 are our features that then map to a specific species. We can use this information to train our model to see what features are used to classify flowers.

# Part 2: Split the dataset into train and test

In [37]:
# Take the dataset and split it into our features (X) and label (y)
X = iris_df.iloc[:, :-1]
y = iris_df.iloc[:, -1]
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .2, random_state=42)


# Part 3: Logistic Regression

In [ ]:
# i. Use sklearn to train a LogisticRegression model on the training set
logistic_model = LogisticRegression()
logistic_model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_point = X_test.iloc[0]
sample_point_reshape = sample_point.values.reshape(1, -1)
probabilities = logistic_model.predict_proba(sample_point_reshape)
#First is Setosa, Second is Versicolor, Third is Virginica
print("Probabilities: ",probabilities)
# iii. Report on the score for Logistic regression model, what does the score measure?
logistic_score = logistic_model.score(X_test, y_test)
print("Model Score: ", logistic_score)
#The score measures how effectively  Logistic Regression can predict the species of iris flowers based on the features. A high accuracy suggests that the model performs well on the test data, while a low score may indicate inadequate performance.
#The score measures the accuracy of the predictions, meaning the number of correct predicitions divided by the total number of predictions. This model's score is a 1 which means that it predicted it 100% correctly.

# iv. Extract the coefficents and intercepts for the boundary line(s)
print("Coefficients: ",logistic_model.coef_)
print("Intercept", logistic_model.intercept_)

Probabilities:  [[0.00377554 0.8269881  0.16923636]]
Model Score:  1.0
Coefficients:  [[-0.39634076  0.9523916  -2.37629048 -1.01351596]
 [ 0.51221741 -0.24809347 -0.21455322 -0.76203024]
 [-0.11587664 -0.70429814  2.5908437   1.7755462 ]]
Intercept [  9.07176597   1.81821925 -10.88998522]


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


# Part 4: Support Vector Machine

In [ ]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svc_model = SVC(probability=True)
svc_model.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
sample_point = X_test.iloc[0]
sample_point_reshaped = sample_point.values.reshape(1, -1)
probabilities = svc_model.predict_proba(sample_point_reshaped)
print("Probabilities: ", probabilities)
# iii. Report on the score for the SVM, what does the score measure?
svc_score = svc_model.score(X_test, y_test)
print("Model Score: ", svc_score)
#The score measures how effectively the SVC can predict the species of iris flowers based on the features. A high accuracy suggests that the model performs well on the test data, while a low score may indicate inadequate performance.
#The score measures the accuracy of the predictions, meaning the number of correct predicitions divided by the total number of predictions. This model's score is a 1 which means that it predicted it 100% correctly.

Probabilities:  [[0.00749132 0.90097531 0.09153337]]
Model Score:  1.0


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  warnings.warn(


# Part 5: Neural Network

In [ ]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlp_model = MLPClassifier()
mlp_model.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
sample_point = X_test.iloc[0]
sample_point_reshape = sample_point.values.reshape(1, -1)
probabilities = mlp_model.predict_proba(sample_point_reshape)
print("Probabilities: ", probabilities)

# iii. Report on the score for the Neural Network, what does the score measure?
mlp_score = mlp_model.score(X_test, y_test)
print("Model Score: ", mlp_score)
#The score measures how effectively the NN  can predict the species of iris flowers based on the features. A high accuracy suggests that the model performs well on the test data, while a low score may indicate inadequate performance.
#The score measures the accuracy of the predictions, meaning the number of correct predicitions divided by the total number of predictions. This model's score is a 1 which means that it predicted it 100% correctly.

# iv: Experiment with different options for the neural network, report on your best configuration
from sklearn.model_selection import GridSearchCV
param_grid = {
    'hidden_layer_sizes': [(5,), (10,), (5, 5), (10, 5)],  # Different layer configurations
    'activation': ['relu', 'tanh', 'logistic'],  # Different activation functions
    'alpha': [0.0001, 0.001, 0.01],  # Different regularization strengths
    'max_iter': [500, 1000]  # Number of iterations
}

grid_search = GridSearchCV(MLPClassifier(), param_grid, cv=5, n_jobs=-1) 
grid_search.fit(X_train, y_train)

best_model = grid_search.best_estimator_
best_params = grid_search.best_params_
best_score = grid_search.best_score_


print("Grid Search Results:")
# Display all results
results = grid_search.cv_results_

for mean_score, params in zip(results['mean_test_score'], results['params']):
    print(f"Mean Test Score: {mean_score:.4f} | Parameters: {params}")

print("\nBest configuration:")
print(f"Best parameters: {best_params}")
print(f"Best cross-validated accuracy: {best_score:.4f}")

# Evaluate on the test set
test_score = best_model.score(X_test, y_test)
print(f"Test set accuracy for best model: {test_score:.4f}")


Probabilities:  [[0.01670161 0.71790206 0.26539633]]
Model Score:  1.0


/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but MLPClassifier was fitted with feature names
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_pe

Grid Search Results:
Mean Test Score: 0.8500 | Parameters: {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes': (5,), 'max_iter': 500}
Mean Test Score: 0.9167 | Parameters: {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes': (5,), 'max_iter': 1000}
Mean Test Score: 0.8917 | Parameters: {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes': (10,), 'max_iter': 500}
Mean Test Score: 0.9667 | Parameters: {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes': (10,), 'max_iter': 1000}
Mean Test Score: 0.7167 | Parameters: {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes': (5, 5), 'max_iter': 500}
Mean Test Score: 0.9250 | Parameters: {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes': (5, 5), 'max_iter': 1000}
Mean Test Score: 0.9000 | Parameters: {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes': (10, 5), 'max_iter': 500}
Mean Test Score: 0.9667 | Parameters: {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes': (10, 5

The Best Configuration ended up being utilizing a ReLu activation function. Using an alpha of .001 which is a penalty applied to larger weights to help prevent overfitting.  This means not much regularization was needed to acheive good results. We utilized 2 hidden layers, first one with 10 and the second one with 5. This allowed us to get a cross validated accuracy of 97.5% which means that the model correctly predicted the species of the flowers 97.5% of the time ocress different subsets of training data. We achieved a test accuracy of 96.67% which meant that the model correctly predicted 96.67% of the species.

# Part 6: K-Nearest Neighbors

In [ ]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier

k_neighbors_model = KNeighborsClassifier(n_neighbors=10)
k_neighbors_model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_point = X_test.iloc[0]
sample_point_reshape = sample_point.values.reshape(1, -1)
probabilities = k_neighbors_model.predict_proba(sample_point_reshape)
print("Probabilities: ", probabilities)

# iii. Report on the score for kNN, what does the score measure?
knn_score = k_neighbors_model.score(X_test, y_test)
print("Model Score: ", knn_score)
#The score measures how effectively the k-NN model can predict the species of iris flowers based on the features. A high accuracy suggests that the model performs well on the test data, while a low score may indicate inadequate performance.
#The score measures the accuracy of the predictions, meaning the number of correct predicitions divided by the total number of predictions. This model's score is a 1 which means that it predicted it 100% correctly.

Probabilities:  [[0.  0.8 0.2]]
Model Score:  1.0


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  warnings.warn(


# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

Since the data set is pretty small, all of the models had a score of 1. Even changing the random state did not change it, and there were only slight changes when the test percentage was closer to the 90% range. The NN with the "best" parameters had an accuracy with less than 1, which could probably be due to the amount of iterations. Something else surprising was that the K-nn model had a score of 1 even for 1 neighbor which must mean the data has little to no outliers, or the flowers tend to stay close to the distribution of their features and not vary much. Maybe a larger dataset would lower the model accuracy. Otherwise I think that all of the models performed well due to the smaller size. Because of this the "best" model is probably the least computational expensive one.